go/jsonclient/client.go

```diff
@@ -0,0 +1,191 @@
+package jsonclient
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	ct "github.com/google/certificate-transparency/go"
+	"golang.org/x/net/context"
+	"golang.org/x/net/context/ctxhttp"
+)
+
+// JSONClient provides common functionality for interacting with a JSON server
+// that uses cryptographic signatures.
+type JSONClient struct {
+	uri        string                // the base URI of the server. e.g. http://ct.googleapis/pilot
+	httpClient *http.Client          // used to interact with the server via HTTP
+	Verifier   *ct.SignatureVerifier // nil if no public key for server available
+}
+
+// New constructs a new JSONClient instance, for the given base URI, using the
+// given http.Client object (if provided).  If a PEM encoded public key is
+// provided, the JSONClient instance will be able to check signatures on signed
+// responses from the server.
+func New(uri string, hc *http.Client, pemEncodedKey string) (*JSONClient, error) {
+	var verifier *ct.SignatureVerifier
+	if pemEncodedKey != "" {
+		pubkey, _, rest, err := ct.PublicKeyFromPEM([]byte(pemEncodedKey))
+		if err != nil {
+			return nil, err
+		}
+		if len(rest) > 0 {
+			return nil, errors.New("extra data found after PEM key decoded")
+		}
+
+		verifier, err = ct.NewSignatureVerifier(pubkey)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if hc == nil {
+		hc = new(http.Client)
+	}
+	return &JSONClient{uri: strings.TrimRight(uri, "/"), httpClient: hc, Verifier: verifier}, nil
+}
+
+// GetAndParse makes a HTTP GET call to the given path, and attempt to parse
+// the response as a JSON representation of the rsp structure.  The provided
+// context is used to control the HTTP call.
+func (c *JSONClient) GetAndParse(ctx context.Context, path string, params map[string]string, rsp interface{}) (*http.Response, error) {
+	if ctx == nil {
```

Comment:
Yep, it's an interface.  The existing `client/logclient.go` code explicitly uses a `nil` context in some places, so I wanted to make that barf here to try to nudge things towards a context-everywhere world.